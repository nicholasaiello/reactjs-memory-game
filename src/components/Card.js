import React, { Component } from 'react';

class Card extends Component {

  constructor(props) {
    super(props);
    this.state = this.initialState;
  }

  get initialState() {
    return { open: false, matched: false };
  }

  get value () { return this.props.value }

  toggleOpen = () => {
    this.setState({ open: !this.state.open });
  }

  markAsMatched = () => {
    this.setState({ matched: true });
  }

  handleClick = () => {
    this.toggleOpen();

    // comes from parent component
    this.props.onClick(this);
  }

  render() {

    let body = '',
      className = 'card',
      onClick = () => { this.handleClick() };
    
    if (this.state.open) {
      body = (<div>{this.props.value}</div>);
      className += ' open' + (this.state.matched ? ' matched' : '');
      onClick = () => {};
    }

    return (
      <div className={className} onClick={onClick}>
        {body}
      </div>
    );

  }

};

Card.defaultProps = {
  value: -1,
  index: -1,
  onClick: () => {}
};

export default Card;
