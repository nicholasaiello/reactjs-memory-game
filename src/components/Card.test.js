import React from 'react';
import renderer from 'react-test-renderer';

import Card from './Card';

describe('Card object', () => {

  it('card facing down', () => {
    const component = renderer.create(
      <Card value={7} />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('card facing up', () => {
    const component = renderer.create(
      <Card value={7} open={true} />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('card matched', () => {
    const component = renderer.create(
      <Card value={7} open={true} matched={true} />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
