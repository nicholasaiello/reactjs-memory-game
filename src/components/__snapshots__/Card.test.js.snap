// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card object card facing down 1`] = `
<figure
  className="card"
  onClick={[Function]}
>
  
</figure>
`;

exports[`Card object card facing up 1`] = `
<figure
  className="card open"
  onClick={[Function]}
>
  <div>
    7
  </div>
</figure>
`;

exports[`Card object card matched 1`] = `
<figure
  className="card open matched"
  onClick={[Function]}
>
  <div>
    7
  </div>
</figure>
`;
